### PR TITLE
feat(progress): add parts for more design customization

### LINKED ISSUE #20062: feat: Add Progress bar radius property
# Feature Request

**Ionic version:**
[x] **4.x**

**Describe the Feature Request**
The ion progress bar component is not easy to customize the border radius of the inner progress bar. We can easily modify the outer progress bar radius, but not the inner one.

**Describe Preferred Solution**
Be able to adjust the inner radius with CSS Properties

**Describe Alternatives**
Would be awesome to have a CSS Property like --inner-progress-radius to be able to adjust the radius easily

### LINKED ISSUE #21820: bug: ion-progress-bar transition cannot be overridden
# Bug Report

**Ionic version:**

[ ] **4.x**
[x] **5.x**

**Current behavior:**
When having quick data value updates (< 150ms), the ion-progress-bar components "jumps" back since the transition of the progress element has not finished yet. Since no CSS vars are used for that and there is no chance to use slots for injecting a progress element, that you can apply custom CSS rules to, there is no possibility to change the `transition-duration`.

[Please see this video](https://www.dropbox.com/s/7dfnvttecyene1k/ion-progress-transition-glitch.mp4?dl=0) (dropbox link) for a demonstration.

**Expected behavior:**
Possibility to change `transition-duration` for ion-progress-bar via CSS vars.

**Steps to reproduce:**
1. Add an `ion-progress-bar` component to a template.
1. Bind the `data` attribute to a local property.
1. Create a loop (e.g. via setInterval) with a short update interval (< 150ms) to manipulate the bound property.
1. See the `ion-progress-bar` component tremble. 

**Repository with demo/reproduction code:**

https://github.com/JZubero/ion-progress-bar-Issue-21820

**Ionic info:**

```
Ionic:

   Ionic CLI                     : 6.10.1 (/home/js/.nvm/versions/node/v10.15.0/lib/node_modules/@ionic/cli)
   Ionic Framework               : @ionic/angular 5.3.0
   @angular-devkit/build-angular : 0.803.28
   @angular-devkit/schematics    : 8.3.28
   @angular/cli                  : 8.3.28
   @ionic/angular-toolkit        : 2.2.0

Capacitor:

   Capacitor CLI   : 2.2.0
   @capacitor/core : 2.2.0

Utility:

   cordova-res (update available: 0.15.1) : 0.14.0
   native-run                             : 1.0.0

System:

   NodeJS : v10.15.0 (/home/js/.nvm/versions/node/v10.15.0/bin/node)
   npm    : 6.14.6
   OS     : Linux 5.4
```

### PR DESCRIPTION
**Pull request by brandyscarney**
## Features

This adds the following parts to progress bar:

| Name | Description |
| ---| ---|
| `progress` | The progress bar that shows the current value (determinate) or slides back and forth (indeterminate). |
| `stream` | The animated circles that appear while buffering, only shows when `buffer` is set and `type` is `"determinate"`. |
| `track` | The track bar is behind the progress bar, if the `buffer` property is set (determinate) the track will be the width of the `buffer` value. |

## Bug fixes

fixes
- setting any of the color variables `--progress-background`, `--background`, etc. no longer override the colors set by the `color` property

resolves #20062 fixes #21820

## Demo

Added additional examples to http://localhost:3333/src/components/progress-bar/test/standalone

I modified the progress bar so that you can set the background of the indeterminate using the `track` part and this also applies to the determinate track / buffer bar, which are the same color. The `progress` part will apply to the current value in determinate, and the moving bar for indeterminate. The `stream` part is only relevant for determinate when the `buffer` value is set. 

progress-parts

### CODE REVIEW

File: core/src/components/progress-bar/readme.md [hunk 1]
```diff
@@ -1,20 +1,18 @@
...
 
 ## Progress Type
 
 ### Determinate
 
-If the percentage of an operation is known, you should use the determinate type. This is the default type and the progress is represented by the `value` property.
+Determinate is the default type. It should be used when the percentage of an operation is known. The progress is represented by setting the `value` property. This can be used to show the progress increasing from 0 to 100% of the track.
 
-A buffer shows circles as animation to indicate some activity. If the `buffer` property is smaller than 1 you can show the additional buffering progress.
+If the `buffer` property is set, a buffer stream will show with animated circles to indicate activity. The value of the `buffer` property will also be represented by how much visible track there is. If the value of `buffer` is less than the `value` property, there will be no visible track. If `buffer`` is equal to `1` then the buffer stream will be hidden.
```
Comment (liamdebeasi) [inline core/src/components/progress-bar/readme.md:11]:
```suggestion
If the `buffer` property is set, a buffer stream will show with animated circles to indicate activity. The value of the `buffer` property will also be represented by how much visible track there is. If the value of `buffer` is less than the `value` property, there will be no visible track. If `buffer` is equal to `1` then the buffer stream will be hidden.
```

File: core/src/components/progress-bar/readme.md [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
 
 ### Indeterminate
 
-If an operation is in progress and it's not necessary to indicate how long it will take.
-
-If you add `reversed="true"`, you receive a query which is used to indicate pre-loading.
+The indeterminate type should be used when it is unknown how long the process will take. The progress bar is not tied to the `value`, instead it continually grows and shrinks along the track until the process is complete.
```
Comment (liamdebeasi) [inline core/src/components/progress-bar/readme.md:15]:
```suggestion
The indeterminate type should be used when it is unknown how long the process will take. The progress bar is not tied to the `value`, instead it continually slides along the track until the process is complete.
```

File: core/src/components/progress-bar/progress-bar.tsx [hunk 3]
```diff
@@ -8,6 +8,11 @@ import { createColorClasses } from '../../utils/theme';
 
 /**
  * @virtualProp {"ios" | "md"} mode - The mode determines which platform styles to use.
+ *
+ * @part progress - The progress bar that shows the current value (determinate) or slides back and forth (indeterminate).
+ * @part stream - The animated circles that appear while buffering, only shows when `buffer` is set and `type` is `"determinate"`.
+ * @part track - The track bar is behind the progress bar, if the `buffer` property is set (determinate) the track will be the
```
Comment (liamdebeasi) [inline core/src/components/progress-bar/progress-bar.tsx:14]:
```suggestion
 * @part track - The track bar is behind the progress bar, if the `buffer` property is set and `type` is `"determinate"` the track will be the
```

File: core/src/components/progress-bar/progress-bar.tsx [hunk 4]
```diff
@@ -8,6 +8,11 @@ import { createColorClasses } from '../../utils/theme';
 
 /**
  * @virtualProp {"ios" | "md"} mode - The mode determines which platform styles to use.
+ *
+ * @part progress - The progress bar that shows the current value (determinate) or slides back and forth (indeterminate).
```
Comment (liamdebeasi) [inline core/src/components/progress-bar/progress-bar.tsx:12]:
```suggestion
 * @part progress - The progress bar that shows the current value when `type` is `"determinate"` and slides back and forth when `type` is `"indeterminate"`.
```

**Review by liamdebeasi — COMMENTED** [review]
The progress animation seems to be broken on both iOS and MD 

https://user-images.githubusercontent.com/2721089/109862982-173b4700-7c2f-11eb-8315-3a199a9b8fc9.mov

**Comment by brandyscarney** [comment]
@liamdebeasi Resolved issue with animation

File: core/src/components/progress-bar/progress-bar.tsx [hunk 5]
```diff
@@ -8,6 +8,11 @@ import { createColorClasses } from '../../utils/theme';
 
 /**
  * @virtualProp {"ios" | "md"} mode - The mode determines which platform styles to use.
+ *
+ * @part progress - The progress bar that shows the current value when `type` is `"determinate"` and slides back and forth when `type` is `"indeterminate"`.
+ * @part stream - The animated circles that appear while buffering, only shows when `buffer` is set and `type` is `"determinate"`.
```
Comment (liamdebeasi) [inline core/src/components/progress-bar/progress-bar.tsx:13]:
```suggestion
 * @part stream - The animated circles that appear while buffering. This only shows when `buffer` is set and `type` is `"determinate"`.
```

File: core/src/components/progress-bar/progress-bar.tsx [hunk 6]
```diff
@@ -8,6 +8,11 @@ import { createColorClasses } from '../../utils/theme';
...
+ * @part track - The track bar is behind the progress bar, if the `buffer` property is set and `type` is `"determinate"` the track will be the
+ * width of the `buffer` value.
```
Comment (liamdebeasi) [inline core/src/components/progress-bar/progress-bar.tsx:15]:
```suggestion
 * @part track - The track bar behind the progress bar. If the `buffer` property is set and `type` is `"determinate"` the track will be the
 * width of the `buffer` value.
```

**Review by liamdebeasi — APPROVED** [review]
A few description tweaks, but otherwise this is good to go. Nice job!

**Pull request merged**